Task7 vizu

frareb opened the pull request:
Dans cette tâche 7 il faudrait des graphiques avec :

1) sélection de l'institut, du gateway, puis du capteur affiché sous forme de ligne ou boxplot (c'est l'objet de cette PR).

2) sélection de l'institut, du gateway, puis de plusieurs capteurs (eg température de l'air, du sol, à 10cm de profondeur) affichés sous forme de lignes (comparaison entre capteurs d'un même gateway).

3) sélection d'un pays ou d'un institut, puis d'un type de capteur (eg humidité relative), affichés sous forme de lignes (comparaison entre gateways d'un même institut ou d'un même pays d'un type de capteur).

Cette PR propose un formulaire pour afficher les données d'un capteur avec des select boxes : institut puis filtrage sur les gateways puis sur les capteurs. 

// Problème #1 : la liste des capteurs disponibles est limitée à
// l'affichage dans /gateways/:gatewayId (10), donc la liste est
// incomplète. Même chose pour la liste des gateways. 
// Dans /controllers/details.js on peut mettre
// props.limit = 30 lignes 12 et 15 pour avoir jusqu'à 30
// capteurs ?

// Problème #2 : il y a un décalage dans les timestamps de deux
// heures (UTC versus paramètres locaux ?).

// Problème #3 : la fonction de moyenne mobile fait une moyenne
// sur un nombre de points plutôt que sur une période de temps
// ce qui posera problème si deux capteurs ont des données à des
// pas de temps différents. Ce n'est pas important pour le moment.

frareb committed:
plotly line chart

frareb committed:
html file for plotly line chart

frareb committed:
plotly from cdn

frareb committed:
urlBase for localhost

frareb committed:
colors and comments

frareb committed:
task7 in corresponding folder

frareb committed:
lineChart responsive

frareb committed:
boxplot html

frareb committed:
boxplot / day

frareb committed:
editable and download

frareb committed:
index to see charts

frareb committed:
Rename boxplotChartPlotly.htlm to boxplotChartPlotly.html

frareb committed:
Rename boxPlotPlotly.js to boxplotChartPlotly.js

StaloneLab commented:
A titre d'information : j'ai fait un `rebase` de la PR sur master afin de la faire fonctionner.

StaloneLab committed:
Allow specific filters for some endpoints

StaloneLab committed:
Artificially fix timestamps

StaloneLab committed:
Allow bulk deletion of data

StaloneLab commented:
Fait : 

- résolution du problème de la limite sur les gateways et capteurs ;
- résolution du problème d'import des données avec des fuseaux horaires différents ;
- ajout de la possibilité de supprimer des données entre deux dates (voir `/sensors/x/datas` en DELETE dans le Swagger).

StaloneLab commented on Task07/boxplotChart/boxplotChartPlotly.html:
```diff
@@ -0,0 +1,26 @@
+<head>
```
Il manque le Doctype et la balise `<html>`

frareb replied:
ok

StaloneLab commented on Task07/boxplotChart/boxplotChartPlotly.js:
```diff
@@ -0,0 +1,221 @@
+  let i = parseInt (c);
+  if (i == 0 || isNaN (c))
+    return "00";
+  i = Math.round (Math.min (Math.max (0, i), 255));
+  return s.charAt ((i - i % 16) / 16) + s.charAt (i % 16);
+}
+/* Convert an RGB triplet to a hex string */
+function convertToHex (rgb) {
+  return hex(rgb[0]) + hex(rgb[1]) + hex(rgb[2]);
+}
+/* Remove '#' in color hex string */
+function trim (s) { return (s.charAt(0) == '#') ? s.substring(1, 7) : s }
+/* Convert a hex string to an RGB triplet */
+function convertToRGB (hex) {
+  let color = [];
+  color[0] = parseInt ((trim(hex)).substring (0, 2), 16);
+  color[1] = parseInt ((trim(hex)).substring (2, 4), 16);
+  color[2] = parseInt ((trim(hex)).substring (4, 6), 16);
+  return color;
+}
+function generateColor(colorStart,colorEnd,colorCount){
+	// The beginning of your gradient
+	let start = convertToRGB (colorStart);
+	// The end of your gradient
+	let end   = convertToRGB (colorEnd);
+	// The number of colors to compute
+	let len = colorCount;
+	//Alpha blending amount
+	let alpha = 0.0;
+	let saida = [];
+	for (let i = 0; i < len; i++) {
+		var c = [];
+		alpha += (1.0/len);
+		c[0] = start[0] * alpha + (1 - alpha) * end[0];
+		c[1] = start[1] * alpha + (1 - alpha) * end[1];
+		c[2] = start[2] * alpha + (1 - alpha) * end[2];
+		saida.push(convertToHex (c));
+	}
+	return saida;
+}
```
Toutes ces fonction concernent les couleurs, elles peuvent sans doute être dans un fichier spécifique `colors.js`.

frareb replied:
Pas encore fait.

StaloneLab commented on Task07/boxplotChart/boxplotChartPlotly.js:
```diff
@@ -0,0 +1,221 @@
+	return saida;
+}
+
+function makeSimpleBoxplotSensor(sensorId=1, nameVar="xxx", nameInst="yyy", nameGateway="") {
```
Les paramètres par défaut ne sont pas obligatoires en JavaScript, inutile de mettre de fausses valeurs si elles ne sont pas valides (xxx, yyy, ...).

frareb replied:
ok

StaloneLab commented on Task07/boxplotChart/boxplotChartPlotly.js:
```diff
@@ -0,0 +1,221 @@
+}
+
+function makeSimpleBoxplotSensor(sensorId=1, nameVar="xxx", nameInst="yyy", nameGateway="") {
+  let timestampNow = Date.now();
+  let timestamp30d = timestampNow - (30*24*60*60*1000) // -30 days
+  let myURL = urlBase + '/Sensors/'+ sensorId +'/datas?start=' + timestamp30d + '&end=' + timestampNow;
+  let myX = [];
+  let myY = [];
+  let myTrace = [];
```
Probablement toutes ces variables pourraient être en `const` au lieu de `let` (voir le commentaire global).

frareb replied:
oui

StaloneLab commented on Task07/boxplotChart/boxplotChartPlotly.js:
```diff
@@ -0,0 +1,221 @@
+      const minTemp = data.data.reduce((min, p) => p.value < min ? p.value : min, data.data[0].value);
+      const maxTemp = data.data.reduce((max, p) => p.value > max ? p.value : max, data.data[0].value);
+
+      for (const i of objKeyByDay){
```
`i`, `j`, etc, doivent être évités sauf lorsqu'ils ont une portée réduite : ces noms de variables portent à confusion car ils ne sont pas parlants.

Préférer un nom complet, ou éventuellement la première lettre du nom de variable de l'objet itéré.

frareb replied:
ok

StaloneLab commented on Task07/boxplotChart/boxplotChartPlotly.js:
```diff
@@ -0,0 +1,221 @@
+function makeSimpleBoxplotSensor(sensorId=1, nameVar="xxx", nameInst="yyy", nameGateway="") {
+  let timestampNow = Date.now();
+  let timestamp30d = timestampNow - (30*24*60*60*1000) // -30 days
+  let myURL = urlBase + '/Sensors/'+ sensorId +'/datas?start=' + timestamp30d + '&end=' + timestampNow;
```
Pas d'objection particulière, mais je préfère utiliser les [template literals](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals)

frareb replied:
fait

StaloneLab commented on Task07/boxplotChart/boxplotChartPlotly.js:
```diff
@@ -0,0 +1,221 @@
+        let myY2 = [];
+        for(let j=0; j<byday[i].length; j++){
+          // let x = new Date(byday[i][j].createdAt - 120*60*1000);
+          let x = new Date(byday[i][j].createdAt); //new Date
+          // console.log(x);
+          let date = x.getDate();
+          let month = x.getMonth() + 1;
+          let year = x.getFullYear();
+          // let hour = x.getHours() + 1;
+          // if (hour < 10) {hour = '0' + hour};
+          // let xx = new Date(year + '-' + month + '-' + date + ' ' + hour + ':00:00');
+          let xx = new Date(year + '-' + month + '-' + date + ' 00:00:00');
```
Archaïque à mon avis, mais ça fonctionne.

Je propose : 

```js
const x = new Date(byday[i][j].createdAt);
x.setMonth(x.getMonth() - 1);
```

frareb replied:
Ici l'objectif est d'avoir une seule valeur de date pour tous les timestamps d'un jour. C'est donc : 
const x = new Date(byday[keyDay][j].createdAt);
          x.setHours(0);
          x.setMinutes(0);
          x.setSeconds(0);
Je mets cela dans un prochain commit.

StaloneLab commented on Task07/boxplotChart/boxplotChartPlotly.js:
```diff
@@ -0,0 +1,221 @@
+let urlListInst = urlBase + "/institutes?page_size=100"
+$.getJSON(urlListInst, function(x) {
+  var select = document.getElementById("selectInst");
+  select.options.length = 0; // vider les options
```
Wow! Je n'avais jamais vu ça, mais la spec dit que c'est ok.

frareb replied:
J'ai utilisé cela pour effacer les options lors d'un changement de select box. Il y a sûrement d'autres possibilités.

StaloneLab commented on Task07/index.html:
```diff
@@ -0,0 +1,8 @@
+
```
Doctype + `html` + `head` + `body`

StaloneLab reviewed (commented):
Il faudrait séparer les fichiers en briques plus petites pouvant être réutilisées.

Sur une première version, charger plusieurs fichiers de JavaScript n'est pas problématique, il est possible de les unifier ensuite si nécessaire.
Je proposerais une version basée sur [Webpack](https://webpack.js.org/) dans un second temps, avec un chargement des modules via [import](https://developer.mozilla.org/fr/docs/Web/JavaScript/Reference/Instructions/import)/[export](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/export).

Cela permettra en plus de supporter les anciens navigateurs sans efforts en utilisant [Babel](https://babeljs.io/).

Sur l'utilisation de `let` et `const`, il y a beaucoup d'approximations qui circulent. Pour résumer, car il y a beaucoup de confusions dans les fichiers : 

- `var` est obsolète car sa portée est mal définie. Il n'y a plus de raisons de l'utiliser aujourd'hui, sauf pour la compatibilité ;
- `const` doit être le remplacement de `var` dans 90% des cas, il désigne une référence fixe (mais la valeur peut changer) ;
- `let` peut être utilisé ponctuellement lorsque la référence doit changer ;
- `let` ET `const` ont une portée limitée, pas seulement `let` : 

> let and const declarations define variables that are scoped to the running execution context's LexicalEnvironment.
Source : ECMA-262 13.3.1 - Let and Const Declarations

Ces histoires de références sont souvent oubliées dans les langages de haut-niveau, mais elles restent importantes ; voici deux exemples de codes, un valide et l'autre invalide : 

Correct : 

```js
// Déclaration d'un objet
const obj = {};

// Allocation dynamique DANS l'objet, la référence n'est donc pas changée
obj.a = "a";
```

Incorrect : 

```js
// Déclaration d'un objet
const obj = {};

// Redéfinition DE l'objet, la constante doit changer sa référence, ce n'est pas possible, il faut donc utiliser un `let`
obj = {a: "a"};
```

frareb committed:
modifs sur PR6 : 
- const/let
- doctype html head body
- pas de valeur par défault sur les fonctions
- i,j -> noms complets
- template literals

StaloneLab commented:
Hum, j'ai l'impression qu'il y a un problème de Git, les premiers commits apparaissent deux fois, ce qui fait qu'il y a des doublons dans les fichiers.

En cas de problèmes pour la résolution, ma branche locale n'est pas à jour avec cette PR, on peut donc sans doute récupérer un historique correct via deux [cherry-pick](https://git-scm.com/docs/git-cherry-pick) ; dites-moi si vous ne parvenez pas à corriger de votre côté. Il faudrait que vous n'ayez pas de modifications locales avant de faire l'opération, car tout l'historique va bouger.

frareb commented:
Je vois bien les commits en double, je ne comprends pas ce qui a pu se passer (depuis ta review j'ai changé de distribution pour Lubuntu 20.04 mais je ne vois pas de rapport j'ai pu tout récupérer sans incident)... si je fais un pull sur cette branche je récupère bien une version à jour. Je n'ai pas de modifications locales. Si tu penses pouvoir résoudre ce problème avec cherry-pick merci de ton aide. Tiens moi au courant et nous pourrons en reparler demain lors de notre point.
